crates/rolldown_utils/src/percent_encoding.rs

```diff
@@ -1,37 +1,40 @@
 /// Authored by @ikkz and adapted by @7086cmd.
-use std::fmt::Write;
 use std::str;
 
+const HEX: [char; 16] =
+  ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'];
 // adapted from "https://github.com/evanw/esbuild/blob/67cbf87a4909d87a902ca8c3b69ab5330defab0a/scripts/dataurl-escapes.html" for how this was derived
 pub fn encode_as_percent_escaped(buf: &[u8]) -> Option<String> {
-  if let Ok(text) = str::from_utf8(buf) {
-    let mut url = String::with_capacity(text.len() * 3);
-    let chars = text.chars().collect::<Vec<_>>();
-    let mut trailing_start = chars.len();
-    while trailing_start > 0 {
-      let c = chars[trailing_start - 1];
-      if c > 0x20 as char || matches!(c, '\t' | '\n' | '\r') {
-        break;
+  str::from_utf8(buf)
+    .map(|text| {
+      let mut url = String::with_capacity(text.len() * 3);
+      let chars = text.chars().collect::<Vec<_>>();
+      let mut trailing_start = chars.len();
+      while trailing_start > 0 {
+        let c = chars[trailing_start - 1];
+        if c > 0x20 as char || matches!(c, '\t' | '\n' | '\r') {
+          break;
+        }
+        trailing_start -= 1;
       }
-      trailing_start -= 1;
-    }
-    for (i, &c) in chars.iter().enumerate() {
-      if matches!(c, '\t' | '\n' | '\r' | '#')
-        || i >= trailing_start
-        || (c == '%'
-          && i + 2 < chars.len()
-          && chars[i + 1].is_ascii_hexdigit()
-          && chars[i + 2].is_ascii_hexdigit())
-      {
-        write!(url, "%{:02X}", c as u32).unwrap();
-      } else {
-        url.push(c);
+      for (i, &c) in chars.iter().enumerate() {
+        if matches!(c, '\t' | '\n' | '\r' | '#')
+          || i >= trailing_start
+          || (c == '%'
+            && i + 2 < chars.len()
+            && chars[i + 1].is_ascii_hexdigit()
+            && chars[i + 2].is_ascii_hexdigit())
+        {
+          url.push('%');
+          url.push(HEX[c as usize >> 4]);
+          url.push(HEX[c as usize & 15]);
```

Comment:
I don't change code other than these three lines, for other changes, I use a combinator to refactor the code.